codec/src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,136 @@
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public final class BrotliDecoder extends ByteToMessageDecoder {
+
+    private enum State {
+        DONE, NEEDS_MORE_INPUT, ERROR
+    }
+
+    private DecoderJNI.Wrapper decoder;
+    private final AtomicBoolean destroyed = new AtomicBoolean();
+
+    static {
+        try {
+            Brotli.ensureAvailability();
+        } catch (Throwable throwable) {
+            throw new ExceptionInInitializerError(throwable);
+        }
+    }
+
+    private State decompress(ByteBuf input, List<Object> output) {
+        for (;;) {
+            switch (decoder.getStatus()) {
+                case DONE:
+                    return State.DONE;
+
+                case OK:
+                    decoder.push(0);
+                    break;
+
+                case NEEDS_MORE_INPUT:
+                    if (decoder.hasOutput()) {
+                        output.add(Unpooled.wrappedBuffer(decoder.pull()));
+                    }
+
+                    if (!input.isReadable()) {
+                        return State.NEEDS_MORE_INPUT;
+                    }
+
+                    ByteBuffer decoderInputBuffer = decoder.getInputBuffer();
+                    decoderInputBuffer.clear();
+                    int readBytes = readBytes(input, decoderInputBuffer);
+                    decoder.push(readBytes);
+                    break;
+
+                case NEEDS_MORE_OUTPUT:
+                    output.add(Unpooled.wrappedBuffer(decoder.pull()));
+                    break;
+
+                default:
+                    return State.ERROR;
+            }
+        }
+    }
+
+    private static int readBytes(ByteBuf in, ByteBuffer dest) {
+        int limit = Math.min(in.readableBytes(), dest.remaining());
+        ByteBuffer slice = dest.slice();
+        slice.limit(limit);
+        in.readBytes(slice);
+        dest.position(dest.position() + limit);
+        return limit;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        decoder = new DecoderJNI.Wrapper(8 * 1024);
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        if (!in.isReadable()) {
+            return;
+        }
+
+        try {
+            State state = decompress(in, out);
+            if (state == State.DONE) {
+                destroy();
+            } else if (state == State.ERROR) {
+                throw new DecompressionException("Brotli stream corrupted");
+            }
+        } catch (Exception e) {
+            destroy();
+            throw e;
+        }
+    }
+
+    private void destroy() {
+        if (destroyed.compareAndSet(false, true)) {
```

Comment:
Does it have to be protected if it always invoked from a single event loop thread?